### DISCUSSION TITLE
How to transform server framework exceptions into TRPC?

### DISCUSSION THREAD

**Discussion by realprabs** [trpc/trpc#3167] [Q&A]
I am using TRPC in AdonisJS, and wanted to transform exceptions thrown by the framework, such as [AuthenticationException](https://github.com/adonisjs/auth/blob/develop/src/Exceptions/AuthenticationException.ts), [InvalidCredentialsException](https://github.com/adonisjs/auth/blob/develop/src/Exceptions/InvalidCredentialsException.ts), [AuthorizationException](https://github.com/adonisjs/bouncer/blob/develop/src/Exceptions/AuthorizationException.ts).

Currently I am doing it by returning objects in `errorFormatter`.
```ts
const t = initTRPC.create({
	errorFormatter(options) {
		const { error, shape, path } = options;
		if (error.cause instanceof InvalidCredentialsException) {
			return {
				message: 'Invalid credentials',
				code: TRPC_ERROR_CODES_BY_KEY.UNAUTHORIZED,
				data: {
					code: 'UNAUTHORIZED',
					httpStatus: 401,
					path,
				},
			};
		}
// Other exceptions
		return shape;
})
```
There are many exceptions and it's fairly a fairly long function now. Is it possible to simplify it? I would like to do something like this:
```ts
import { defaultFormatter } from '@trpc/server/dist/error/formatter';

const t = initTRPC.create({
	errorFormatter(options) {
		const { error, shape, path } = options;
		if (error.cause instanceof InvalidCredentialsException) {
			return defaultFormatter({
				...options,
				error: new TRPCError({ code: 'UNAUTHORIZED', message: 'Invalid credentials' }),
			});
		}
		return shape;
})
```
That throws `Error: Package subpath './dist/error/formatter' is not defined by "exports"`.

It would be great to have a standard way of converting any exception into the "right" TRPC exception would be good. By default all those errors result in an `Internal server error` TRPC error.

**Comment by KATT** [comment]
I'm unfamiliar with adonis, but here's an idea:

```ts
// do all your mapping in 1 distinct function
function getAdonisException(cause: unknown) {
  // maybe there's a base exception here?
  if (!(cause instanceof AdonisException)) {
    return null;
  }
  // maybe there's a discriminated union on the Adonis exceptions to make the mapping easier?
  // otherwise, do a bunch of instance of here.
  
  
}
export const t = initTRPC.context<Context>().create({
  errorFormatter({ shape, error }) {
    return {
      ...shape,
      data: {
        ...shape.data,
        adonis: getAdonisException(error.cause),
      },
    };
  }
})
```

**Reply by realprabs** [reply]
This code does not change the output TRPC error. It returns an internal server error.

The question isn't really specific to Adonis. Any framework throwing an error (Model not found, Unauthorized etc.) results in a `TRPCError({code: "INTERNAL_SERVER_ERROR"})` unless we catch it in every procedure. I am trying to avoid that. When an Unauthorized error is thrown by the framework, I want to return a `TRPCError({code: 'UNAUTHORIZED'})` to the client. The current code I have above already works, but looks ugly.

It would be simpler if `import { defaultFormatter } from '@trpc/server/dist/error/formatter';` was allowed.

**Comment by aa-ndrej** [comment]
@jaysson did you manage to find a good solution?

Right now I am setting `shape.data.httpStatus` and then manually mapping the HTTP status codes to tRPC error codes and then setting `shape.data.code`. This works for me and also changes the HTTP status code of the response.

~~~ts
import { Exception } from '@adonisjs/core/build/standalone'
import { initTRPC, TRPCError } from '@trpc/server'
import { Context } from 'App/Trpc/router'


const httpToTrpc: Record<number, TRPCError['code'] | undefined> = {
	400: 'BAD_REQUEST',
	401: 'UNAUTHORIZED',
	403: 'FORBIDDEN',
	404: 'NOT_FOUND',
	405: 'METHOD_NOT_SUPPORTED',
	408: 'TIMEOUT',
	409: 'CONFLICT',
	412: 'PRECONDITION_FAILED',
	413: 'PAYLOAD_TOO_LARGE',
	422: 'UNPROCESSABLE_CONTENT',
	429: 'TOO_MANY_REQUESTS',
	499: 'CLIENT_CLOSED_REQUEST',
	500: 'INTERNAL_SERVER_ERROR',
}
function getTrpcCode(httpCode: number): TRPCError['code'] {
	const trpcCode = httpToTrpc[httpCode]
	if (trpcCode) return trpcCode
	// else:

	if (httpCode >= 400 && httpCode < 500) return 'BAD_REQUEST'
	else return 'INTERNAL_SERVER_ERROR'
}


const t = initTRPC.context<Context>().create({
	errorFormatter({ shape, error }) {

		if (error.cause instanceof Exception) {
			shape.data.httpStatus = error.cause.status
			shape.data.code = getTrpcCode(error.cause.status)
		}
		// Other exceptions.

		return shape
	},
})
~~~
\
I am also interested in what is the _official_ way of intercepting and especially changing errors. In the Docs ([Handling errors](https://trpc.io/docs/server/error-handling#handling-errors) it says that we can use the `onError` method to handle and **change** errors. However when I can not change e.g. `error.code` from this method since it is readonly.
@KATT can you maybe give an example of how to intercept and **change** an error and the corresponding HTTP status code of the response globally? Or my above code already the _official_ way to do it?

**Reply by twonil-jordan-stout** [reply]
I too am trying to solve for this too.

We're using an already built service layer that's using `http-errors` under the hood and I'd like to map those errors to TRPC errors in a global & dynamic way.
